cunumeric/array.py

```diff
@@ -1252,6 +1252,92 @@ def dump(self, file):
     def dumps(self):
         return self.__array__().dumps()
 
+    def fft(self, s, axes, kind, direction, norm):
+        # Dimensions check
+        if self.ndim > 3:
+            raise NotImplementedError(
+                f"{self.ndim}-D arrays are not supported yet"
+            )
+
+        # Type
+        fft_output_type = self.dtype
+        if kind == FFTCode.FFT_D2Z:
+            fft_output_type = np.complex128
+        elif kind == FFTCode.FFT_R2C:
+            fft_output_type = np.complex64
+        elif kind == FFTCode.FFT_Z2D:
+            fft_output_type = np.float64
+        elif kind == FFTCode.FFT_C2R:
+            fft_output_type = np.float32
+
+        # Axes and sizes
+        user_axes  = axes is not None
+        user_sizes = s is not None
+        if not user_axes:
+            fft_axes = list(range(-len(s),0)) if user_sizes else list(range(self.ndim))
+        else:
+            fft_axes = axes
+        fft_axes = [x % self.ndim for x in fft_axes]
+
+        fft_s = list(self.shape)
+        if user_sizes:
+            for idx, ax in enumerate(fft_axes):
+                fft_s[ax] = s[idx]
+
+        # Shape
+        fft_input        = self
+        fft_output_shape = self.shape
+        if user_sizes:
+            # Zero padding if any of the user sizes is larger than input
+            zero_padded_input = self
+            if np.any(np.greater(fft_s, fft_input.shape)):
+                # Create array with superset shape, fill with zeros, and copy input in
+                max_size = tuple(np.maximum(fft_s,fft_input.shape))
+                zero_padded_input = ndarray(shape=max_size, dtype=fft_input.dtype)
+                zero_padded_input.fill(0)
+                zero_padded_input._thunk.set_item(tuple(slice(0,i) for i in fft_input.shape), fft_input._thunk)
+
+            # Slicing according to final shape
+            fft_input_shape = list(fft_input.shape)
+            for idx, ax in enumerate(fft_axes):
+                fft_input_shape[ax] = s[idx]
+            fft_input_shape = tuple(fft_input_shape)
+            fft_input  = ndarray(shape=fft_input_shape, thunk=zero_padded_input._thunk.get_item(tuple(slice(0,i) for i in fft_s))).copy()
+            fft_output_shape = fft_input_shape
+
+        # R2C/C2R require different output shapes
+        if fft_output_type != self.dtype:
+            fft_output_shape = list(fft_output_shape)
+            # R2C/C2R dimension is the last axis
+            if direction == FFTDirection.FORWARD:
+                fft_output_shape[fft_axes[-1]] = fft_output_shape[fft_axes[-1]]//2 + 1
+            else:
+                fft_output_shape[fft_axes[-1]] = s[-1] if user_sizes else 2 * (fft_input.shape[fft_axes[-1]] - 1)
```

Comment:
You're sure about the `2 * (N - 1)` ?
For a "logical" FFT size of `N`, the real size is `N` and the complex `N//2 + 1`. So far so good. 

So you can go from the real (`N`) to complex (`N//2+1`), no problem.

But you can't go from the complex to the real, because `2 * ( (N//2+1) - 1) != N` when `N` is odd. 
Am I missing something?

Reply:
Yup, if you want N to be of odd size in C2R, the output size must be specified. From the numpy docs:

```
**Returns:  out: ndarray**
The truncated or zero-padded input, transformed along the axis indicated by axis, or the last one if axis is not specified. The length of the transformed axis is n, or, if n is not given, 2*(m-1) where m is the length of the transformed axis of the input. To get an odd number of output points, n must be specified.
```

That said, I'll double check to make sure that our tests include odd and even sizes.

Reply:
ACK; didn't know about this.